src/hooks/useSearch.js

```diff
@@ -0,0 +1,166 @@
+import appConfig from 'app.config';
+import { client } from 'client';
+import { getArrayFields, getFields, prepass } from 'gqty';
+import { useCallback, useEffect, useState } from 'react';
+import { useDebounce } from 'use-debounce';
+import { uniqBy } from './usePagination';
+
+const searchInputDebounceMs = 500;
+
+/**
+ * useSearch hook enables a user to perform search functionality from their WordPress site
+ * with proper debouncing of the search input, and pagination via the `loadMore` function.
+ *
+ * @returns {{searchQuery: string, setSearchQuery: (newValue) => void, searchResults: object[] | null, loadMore: () => void, isLoading: boolean, pageInfo: object;}} Result object
+ */
+export default function useSearch() {
+  const [searchQuery, setSearchQuery] = useState('');
+  const [debouncedSearchQuery] = useDebounce(
+    searchQuery,
+    searchInputDebounceMs
+  );
+  const [searchResults, setSearchResults] = useState(null);
+  const [pageInfo, setPageInfo] = useState(null);
+  const [isLoading, setIsLoading] = useState(false);
+
+  /**
+   * Fetch results based on the search query and cursor if we are paginating.
+   * @param {string} searchQuery The user inputted search query
+   * @param {string | undefined} endCursor The end cursor if we are paginating
+   * @returns
+   */
+  async function fetchResults(searchQuery, endCursor = undefined) {
+    /**
+     * Do a prepass request for the search result metadata so we can make a
+     * proper subsequent request for the title and content based on the content
+     * type.
+     *
+     * typically this can be done with the ... on NodeWithTitle interface, but there is
+     * currently a bug.
+     *
+     * @see https://gqty.dev/docs/client/helper-functions#prepass
+     * @see https://github.com/gqty-dev/gqty/issues/733
+     */
+    const metadata = await client.client.resolved(() => {
+      const { nodes, pageInfo } = client.client.query.contentNodes({
+        first: appConfig?.postsPerPage,
+        after: endCursor,
+        where: { search: searchQuery },
+      });
+
+      prepass(nodes, 'databaseId', 'id', 'uri', 'date', '__typename');
```

Comment:
`prepass` does essentially a mapping to resolve those properties. However it certainly looks wierd as it does not return anything. I'm wandering if it's more reasonable to just do `Array.from` and save the result in one new constant:

`const resolvedNodes = Array.from(nodes);`

Reply:
I agree it does look odd, but it is the proper usage. I was also seeing some weird behavior without it, as we are making a subsequent request to get the title and content based on the content type. I've added a comment that describes why we are using it and a link to the [prepass](https://gqty.dev/docs/client/helper-functions#prepass) docs.

Reply:
Even with the `prepass` function here I notice two Graphql requests for the same query. The second one does pass on the extra parameters. I wonder if this is a limit of GQty...

https://user-images.githubusercontent.com/328805/159459434-6de02468-14eb-4e3c-8d8d-f39b1ac702b4.mov

Reply:
I don't think we'll have anyway around that since one query relies on another, resulting in a cascading request. Ideally we would use something like `... on NodeWithTitle` and `... on NodeWIthExcerpt` in the first request, only resulting in one request to get all the data. However, these interfaces are not properly added to the `schema.generated` for some reason. I've opened an issue on GQty:

https://github.com/gqty-dev/gqty/issues/733

Reply:
@theodesp The video you posted shows the UI hanging on the loading animation. Did you have a breakpoint set or is this something that you were experiencing?

Reply:
@theodesp Oh, nvm, the video was having playback issues for me.